src/read.c

```diff
@@ -501,17 +501,18 @@ void ReadCallVarAss (
     if (WarnOnUnboundGlobalsRNam == 0)
       WarnOnUnboundGlobalsRNam = RNamName("WarnOnUnboundGlobals");
 
-    if ( type == 'g'
-      && mode != 'i'
-      && STATE(CountNams) != 0
-      && var != STATE(CurrLHSGVar)
-      && VAL_GVAR(var) == 0
-      && ExprGVar(var) == 0
-      && ! STATE(IntrIgnoring)
-      && ! GlobalComesFromEnclosingForLoop(var)
-      && (GAPInfo == 0 || !IS_REC(GAPInfo) || !ISB_REC(GAPInfo,WarnOnUnboundGlobalsRNam) ||
-             ELM_REC(GAPInfo,WarnOnUnboundGlobalsRNam) != False )
-      && ! SyCompilePlease )
+    if ( type == 'g'                /* Reading a global variable          */
+      && mode != 'i'                /* Not inside 'IsBound'               */
+      && STATE(CountNams) != 0      /* Inside a function                  */
+      && var != STATE(CurrLHSGVar)  /* Not LHS of assignment              */
+      && VAL_GVAR(var) == 0         /* Not an existing global var         */
+      && ExprGVar(var) == 0         /* Or an auto var                     */
+      && ! STATE(IntrIgnoring)      /* Not currently ignoring parsed code */
+      && ! GlobalComesFromEnclosingForLoop(var) /* ? */
```

Comment:
Why the `/* ? */` ?  Doesn't seem to be helpful either way. Perhaps: "Don't warn again for index variable of an enclosing for loop"
An example where this matters:
```
f:=function()
  for j in [1..4] do
    Print(j,"\n"); od;
end;
```
If `j` is not defined, GAP only warns for the first occurrence of `j`. With the `GlobalComesFromEnclosingForLoop` call removed, it warns on both. Which kind of makes sense, because by the time you execute the second line, `j` will no longer be undefined.